js-pkg/client/src/indexeddb.ts

```diff
@@ -0,0 +1,206 @@
+    if (origin === this) {
+      return
+    }
+
+    // We attempt to write in a loop. If we are preempted by another writer, we load the latest
+    // updates and try again.
+    while (true) {
+      await this.loadFromDb()
+      let key = this.updateKey()
+      let newCount = await this.insertValue(key, update)
+
+      if (newCount === null) {
+        // Another writer wrote before we could; reload and try again.
+        continue
+      }
+
+      if (newCount > MAX_UPDATES_IN_STORE) {
+        key = this.updateKey()
+        if (!(await this.saveWholeState(key))) {
+          // Another writer wrote before we could; reload and try again.
+          continue
+        }
+      }
+
+      break
+    }
+
+    this.broadcastChannel.postMessage(this.lastUpdateKey)
+  }
+
+  async getAllValues(range?: IDBKeyRange): Promise<Array<BytesWithKey>> {
+    let transaction = this.db.transaction(OBJECT_STORE_NAME)
+    let objectStore = transaction.objectStore(OBJECT_STORE_NAME)
+    const request = objectStore.getAll(range)
+
+    let result = await new Promise<Array<BytesWithKey>>((resolve, reject) => {
+      request.onsuccess = () => {
+        resolve(request.result)
+      }
+      request.onerror = reject
+    })
+
+    return await Promise.all(
+      result.map(async (data) => {
+        let value = await decryptData(data.value, this.encryptionKey)
+
+        return {
+          key: data.key,
+          value,
+        }
+      }),
+    )
+  }
+
+  async saveWholeState(key: number): Promise<boolean> {
+    const update = Y.encodeStateAsUpdate(this.doc)
+    const encryptedUpdate = await encryptData(update, this.encryptionKey)
+    let transaction = this.db.transaction(OBJECT_STORE_NAME, 'readwrite')
+    let objectStore = transaction.objectStore(OBJECT_STORE_NAME)
+
+    if (await this.hasValue(objectStore, key)) {
+      return false
+    }
+
+    let range = IDBKeyRange.upperBound(key, false)
+    objectStore.delete(range)
+
+    objectStore.add({
+      key,
+      value: encryptedUpdate,
+    })
+
+    return true
+  }
+
+  async hasValue(objectStore: IDBObjectStore, key: number): Promise<boolean> {
+    const request = objectStore.get(key)
+    await new Promise<void>((resolve, reject) => {
+      request.onsuccess = () => resolve()
+      request.onerror = reject
+    })
+
+    return request.result !== undefined
+  }
+
+  /**
+   * Insert a value into IndexedDB. Return the new count of updates in the store if the value was inserted,
+   * or null if the desired key already exists.
+   **/
+  async insertValue(key: number, value: Uint8Array): Promise<number | null> {
+    const encryptedValue = await encryptData(value, this.encryptionKey)
+    let objectStore = this.db
+      .transaction(OBJECT_STORE_NAME, 'readwrite')
+      .objectStore(OBJECT_STORE_NAME)
+
+    if (await this.hasValue(objectStore, key)) {
+      return null
+    }
+
+    const request = objectStore.put({
+      key,
+      value: encryptedValue,
+    })
```

Comment:
I guess it's still possible for two threads to write over each other here (with read-then-write), but I'm guessing that's not a big deal?

Reply:
(discussed offline, the transaction will prevent this)

Reply:
From offline: we think the `readwrite` transaction prevents all other reads from happening. If that's true, the read-then-write wouldn't be an issue.